File: Python/textbook_track/chapter06/ba6h.py
# Implement ColoredEdges
import re
import sys

from enum import Enum


class Sign(Enum):
    PLUS = '+'
    MINUS = '-'


class SignedNumber:
    def __init__(self, sign, number):
        self._sign = sign
        self._number = number

    def __repr__(self):
        return f'{self._sign.value}{self._number}'

    @property
    def sign(self):
        return self._sign

    @property
    def number(self):
        return self._number


class Chromosome:
    def __init__(self, synteny_blocks):
        self._synteny_blocks = tuple(synteny_blocks)

    def __repr__(self):
        return f"({' '.join(map(str, self._synteny_blocks))})"

    def __len__(self):
        return len(self._synteny_blocks)

    def __iter__(self):
        yield from self._synteny_blocks


class Cycle:
    def __init__(self, nodes):
        self._nodes = tuple(nodes)

    def __repr__(self):
        return f"({' '.join(map(str, self._nodes))})"

    def __getitem__(self, k):
        if k == len(self._nodes):
            return self._nodes[0]

        return self._nodes[k]


class Genome:
    def __init__(self, chromosomes):
        self._chromosomes = chromosomes

    def __repr__(self):
        return '\n'.join(map(str, self._chromosomes))

    def __iter__(self):
        yield from self._chromosomes


def read_genome(line):
    genome_parts = re.findall(r'\((.*?)\)', line)
    chromosomes = []
    for part in genome_parts:
        synteny_blocks = []
        for item in part.split():
            signed_number = SignedNumber(Sign(item[0]), int(item[1:]))
            synteny_blocks.append(signed_number)
        chromosomes.append(Chromosome(synteny_blocks))
    return Genome(chromosomes)


def chromosome_to_cycle(chromosome):
    nodes = []
    for block in chromosome:
        if block.sign == Sign.PLUS:
            nodes.append(2*block.number-1)
            nodes.append(2*block.number)
        else:
            nodes.append(2*block.number)
            nodes.append(2*block.number-1)
    return Cycle(nodes)


def calc_colored_edges(genome):
    edges = []
    for chromosome in genome:
        cycle = chromosome_to_cycle(chromosome)
        n = len(chromosome)
        for ix in range(1, n+1):
            edges.append((cycle[2*ix-1], cycle[2*ix]))
    return edges


def main():
    reader = sys.stdin
    genome = read_genome(next(reader))
    result = calc_colored_edges(genome)
    print(', ' .join(f'({a}, {b})' for a, b in result))


if __name__ == '__main__':
    main()

File: Scala/src/main/scala/TextbookTrack/Chapter06/BA6H.scala
package TextbookTrack.Chapter06

object BA6H {
  import scala.util.matching.Regex

  sealed trait Sign
  final case object Plus extends Sign {
    override def toString: String = "+"
  }
  final case object Minus extends Sign {
    override def toString: String = "-"
  }

  object Sign {
    def apply(char: Char): Sign = char match {
      case '+' => Plus
      case '-' => Minus
      case _ => throw new Exception("Unknown sign.")
    }
  }

  final case class SignedNumber(sign: Sign, number: Int) {
    override def toString: String = s"$sign$number"
  }

  final case class Chromosome(syntenyBlocks: Vector[SignedNumber]) {
    override def toString: String = syntenyBlocks.mkString("(", " ", ")")

    def length: Int = syntenyBlocks.length

    def flatMap[T](f: SignedNumber => Vector[T]): Vector[T] = syntenyBlocks.flatMap(f)
  }

  final case class Cycle(nodes: Vector[Int]) {
    override def toString: String = nodes.mkString("(", " ", ")")

    private val length: Int = nodes.length

    def apply(k: Int): Int = {
      if (k == length) nodes(0)
      else nodes(k)
    }
  }

  final case class Genome(chromosomes: List[Chromosome]) {
    override def toString: String = chromosomes.map(_.toString).mkString("\n")

    def flatMap[T](f: Chromosome => List[T]): List[T] = chromosomes.flatMap(f)
  }

  final case class Edge(a: Int, b: Int) {
    override def toString: String = s"($a, $b)"
  }

  def readGenome(line: String): Genome = {
    val pattern: Regex = "\\((.*?)\\)".r
    val chromosomes: List[Chromosome] =
      pattern
        .findAllIn(line)
        .map{ part =>
          val syntenyBlocks: Vector[SignedNumber] =
            part
              .dropRight(1)
              .drop(1)
              .split(" ")
              .map{ item => SignedNumber(Sign(item(0)), item.drop(1).toInt) }
              .toVector
          Chromosome(syntenyBlocks)
        }
        .toList
    Genome(chromosomes)
  }

  private def chromosomeToCycle(chromosome: Chromosome): Cycle = {
    val nodes: Vector[Int] = chromosome.flatMap{
      case SignedNumber(sign, number) =>
        if (sign == Plus) Vector(2*number - 1, 2*number)
        else Vector(2*number, 2*number - 1)
    }
    Cycle(nodes)
  }

  def calcColoredEdges(genome: Genome): List[Edge] = {
    genome.flatMap{ chromosome =>
      val cycle: Cycle = chromosomeToCycle(chromosome)
      val n: Int = chromosome.length
      (1 to n).map(ix => Edge(cycle(2*ix - 1), cycle(2*ix))).toList
    }
  }

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val genome: Genome = readGenome(reader.next())
    val result: List[Edge] = calcColoredEdges(genome)
    println(result.map(_.toString).mkString(", "))
  }
}
